### ISSUE TITLE
Incorrect top output (instantaneous vs interval average)

### ISSUE THREAD

**Issue by txtsd** [brndnmtthws/conky#189]
`top` output seems to show an instantaneous snapshot of CPU usage instead of average CPU usage over the last interval.
Here's a screenshot of my compiling p7zip from source. I've used `make -j2` and you can see both cores are maxed out in the graph. But the processes don't show two ~50% processes. I did see them intermittently, but that's not how CPU usage output should work.
![screenshot](https://i.imgur.com/f3SU08J.png)

Conky version: 1abd253470f5a230b221c2993e54f6137e0ed2b3

**Comment by KamranMackey** [comment]
@txtsd Try updating your conky. That git commit is older than the current master.

**Comment by txtsd** [comment]
@KamranMackey 
None of the commits since then affect this issue and, as expected, the issue still exists.

**Comment by X0rg** [comment]
Hi. Tested with a58a6d4, and I have the same issue: when a process use 4 cores of my CPU, the reported CPU usage is 200% for this process instead 400%, but CPU gauges are OK.
These incorrect values are problematic.

**Comment by X0rg** [comment]
In the same way, when I build something on GNU/Linux:
![conky](https://cloud.githubusercontent.com/assets/2873785/23329796/e04ce740-fb3f-11e6-8978-259b219877ec.png)
The first line of CPU column is `${top cpu 1}`.

**Comment by lasers** [comment]
2 years 8 months passed. Can you determine if you're still having this problem today on `1.10.8` or preferably `1.10.9_pre` (git)? The older versions are not trustworthy due to too many changes that can be hard to track. Additionally, I'd like to see configs too so we know what variables are being used... and if we can reproduce this issue easily. Thank you.

**Comment by lasers** [comment]
@X0rg Okay. Just to be clear... You're referring no more than just `180.27%`? 

I had to knock off few cores off before I can reproduce. I am able to get `303%` for both `${top cpu 1}` and `${top cpu 2}` instead of `100.?`. Apparently, it's not exclusive to `${top cpu 1}`.

Your separators are throwing me off. In the second screenshot, `top` is reporting low numbers, eg `1,03` (`1.03%`) and the cpubars are too high for such numbers. Different screenshot = second issue?

![2018-08-08-185018_348x300_scrot](https://user-images.githubusercontent.com/852504/43870335-99f3afe2-9b3c-11e8-94e9-94da4b2105c8.png)

**Comment by X0rg** [comment]
@lasers `${cpubar}` and `${top}` are not reflecting the same CPU usage:
- `${cpubar}` is showing 100% CPU usage on each core (4x100=**400%**)
- `${top}` is showing less than 400% (180.27+31.46+1+0.5=**213,23%**)

Another screenshot while running pts/c-ray-1.1.0 with Linux **LTS 4.16.60**:
![Screenshot 1](https://user-images.githubusercontent.com/2873785/43887523-665becd4-9bbf-11e8-94fc-52c32aad8b58.png)
On this screenshot, the behavior is normal (395+2+2+1=400%). :heavy_check_mark:

And with Linux **4.17.11 with CK patchset**:
![Screenshot 2](https://user-images.githubusercontent.com/2873785/43887591-99ef23ae-9bbf-11e8-97bb-a88ac32c00c2.png)
On left, `top` is showing 400% for `c-ray-mt` while Conky on right is showing 212.98%. :x:

On txtsd's screenshot, I can see he was using Linux-CK too. In other words, Conky doesn't deal properly with BFS/MuQSS CPU scheduler.

**Comment by su8** [comment]
The sleep during intervals should be one clock tick which in linux is 100ms. Due to the nature of update time interval - the variable it collects cpu usage every x seconds instead.

**Comment by su8** [comment]
@X0rg Can you try the following patch:

```diff
diff --git a/src/linux.cc b/src/linux.cc
index f2844d38..dc4222ab 100644
--- a/src/linux.cc
+++ b/src/linux.cc
@@ -932,6 +932,12 @@ int update_running_processes(void) {
 }
 
 int update_cpu_usage(void) {
+  struct timespec tc = {0L, 500L * 1000000L};
+  update_stat();
+  if (-1 == (nanosleep(&tc, NULL))) {
+    NORM_ERR("update_cpu_usage(): nanosleep failed");
+    return 0;
+  }
   update_stat();
   return 0;
 }
```

To compile against the master branch:

```bash
git clone https://github.com/brndnmtthws/conky
cd conky
mkdir -p build
cd build
cmake ..
make -j4 # 4 cores to run in parallel
```

**Comment by X0rg** [comment]
@su8 Yes. There is no error about `update_cpu_usage()` with this patch.

**Comment by su8** [comment]
Done mate :smiley:

**Comment by X0rg** [comment]
Oh, wait... I'm sorry, I misspoke...
When I said "no error", I wanted to say that:

In the following code
```
if (-1 == (nanosleep(&tc, NULL))) {
  NORM_ERR("update_cpu_usage(): nanosleep() failed");
  return 0;
}
```
`nanosleep` works ("no error" = no `NORM_ERR()` in this case), but it doesn't fix this issue. :confused:

**Comment by awakenedtechnologies** [comment]
Testing with the patch from a few days ago, I see no difference.  But, on the plus side, conky is still awesome.

**Comment by txtsd** [comment]
I don't use conky anymore as I've moved to wayland.

**Comment by Caellian** [comment]
This is expected behavior.

`${top cpu}` and `${cpu}`/`${cpubar}` measure different things, both correctly:
- `calc_cpu_each`: <code>amount = mul * (proc &Delta;jiffies) / (total &Delta;jiffies)</code>; akka per-process share, averaged over the whole `update_interval`
- `update_stat` is aggregate per-core active/total ratio.

`make -j2` spawns many short-lived gcc processes that rotate across cores. Cores stay maxed (cpubar high), but no single process sustains 50% over the interval so `top` shows lower per-process numbers.

Closing as wontfix.
